### PR TITLE
Only try summarizing 3 times consecutively before restarting the summarizer client + refactors

### PR DESCRIPTION
**Pull request by arinwt**
1. First attempt will be normal summary attempt
2. Second attempt will ask the server for what the latest summary ack should be
3. Third attempt will ask server and do full tree with no handles even for unchanged subtrees
4. After 3 consecutive failures, the summarizer client will close itself

Parent client will respawn a new summarizer and start all over.

Also added several refactors in this PR:
- added union typing for Summarizer stop reason codes with comments
- added interface typing for summarize error codes and descriptions
- changed flow of telemetry around a summary attempt (reflected in tests as well)
- removed generateSummaryWithLogging and pulled it up to the summarize function
- changed summarizer closing itself order to better ensure last summary is executed
- more try-catching and reorg around ContainerRuntime.generateSummary to reduce unhandled errors.

### CODE REVIEW

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 1]
```diff
@@ -799,7 +827,7 @@ export class Summarizer extends EventEmitter implements ISummarizer {
             // Raise error to parent container.
             this.emit("summarizingError", createSummarizingWarning("Summarizer: HandleSummaryAckFatalError", true));
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:828]: I believe we either need to eliminate this type of error or remove try/catch from handleSummaryAcks(). I'd vote for the former.

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 2]
```diff
@@ -100,13 +100,26 @@ export interface ISummarizerEvents extends IEvent {
...
+    /** Summarizer client failed to summarize in all 3 consecutive attempts. */
+    | "failToSummarize"
+    /** Highly unexpected error encountered while responding to a summary ack error. */
+    | "handleSummaryAckFatalError"
+    /** Summarizer client detected that its parent is no longer elected the summarizer. */
+    | "parentNoLongerSummarizer"
+    /** Parent client reported that it is no longer connected. */
+    | "parentNotConnected"
+    /** Parent client reported that it is no longer elected the summarizer. */
+    | "parentShouldNotSummarize"
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:113]: Strictly looking at this list, it's not clear what's the difference between these two cases:

/** Summarizer client detected that its parent is no longer elected the summarizer. */ 
| "parentNoLongerSummarizer"

/** Parent client reported that it is no longer elected the summarizer. */
| "parentShouldNotSummarize"

Isn't it one and the same? If not, some clarification would be great!

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 3]
```diff
@@ -465,11 +480,19 @@ export class RunningSummarizer implements IDisposable {
...
+            // On nack or error, try again fetching latest from storage server
+            if (true === await this.summarize(reason, { refreshLatestAck: true, fullTree: false })) {
+                return;
+            }
+            // On another failure, run the full tree
+            if (true === await this.summarize(reason, { refreshLatestAck: true, fullTree: true })) {
+                return;
             }
+            // If all 3 attempts failed, close the summarizer container
+            this.internalsProvider.stop("failToSummarize");
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:495]: Should there be an error event written here? I.e. it should be rare and almost every such case feels like should be a bug, so we should immediately look into it, vs. reporting reason for summarizer closure being general event that we will look mostly in aggregate.

Alternatively, we can make eventName: "StoppingSummarizer" use different categories (including "error" category) for some of the reasons (for now - only this reason). It's not great that events would land in different tables, but we are already in the same boat RE some _cancel events, so I always use unions :)

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 4]
```diff
@@ -486,94 +509,99 @@ export class RunningSummarizer implements IDisposable {
...
-            return undefined;
-        }
+            ++this.summarizeCount;
+
+            // Wait to generate and send summary
+            const summaryData = await this.generateSummaryWithLogging(reason, options);
+            this.heuristics.recordAttempt(summaryData?.referenceSequenceNumber);
+            if (!summaryData || !summaryData.submitted) {
+                // Did not send the summary op
+                this.raiseSummarizingError("Error while generating or submitting summary");
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:529]: Since you are poking at all this code - any chance we can replace summaryData.submitted to summaryData.error (negative) and force layers provide actual error object that goes into telemetry and events?

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 5]
```diff
@@ -486,94 +509,99 @@ export class RunningSummarizer implements IDisposable {
...
+        message: string,
+        options: Omit<IGenerateSummaryOptions, "summaryLogger">,
+    ): Promise<GenerateSummaryData | undefined> {
+        const { refreshLatestAck, fullTree } = options;
         const summarizingEvent = PerformanceEvent.start(this.logger, {
             eventName: "GenerateSummary",
             message,
             timeSinceLastAttempt: Date.now() - this.heuristics.lastAttempted.summaryTime,
             timeSinceLastSummary: Date.now() - this.heuristics.lastAcked.summaryTime,
-            safe: safe || undefined,
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:568]: I'd remove || undefined, I do not think it saves that much in COGSs or bandwidth, but makes it easier to read logs

**Comment by vladsud** [comment]
Love the direction!

Q: Can we easily count (i.e. look at aggregates) how many times we failed/succeeded performing 1-3 steps? I.e. understand the funnel and how much each step actually helps?
Also can we figure out (in aggregate) whether step 2 was actually using different parent (i.e. refresh actually changed our knowledge) or not?

Reason for all these questions - I expect some of the issues to be simply resolved by retry (i.e. storage is down or returns some internal error). It would be great to have a view into the funnel (what step and why it was resolved or not) to assess if we want to make more steps (or simply - more retries) before invoking full summary and eventually giving up.
Full summary will definitely help some cases, but it also could make some issues worse simply due to size (see Jatin's look into stats in issue https://github.com/microsoft/FluidFramework/issues/5690 - we do not know yet all the details, but summary sizes of 29Mb is challenging, it's possible it's due to old doc with inline images, we do not know yet).

**Review by vladsud — APPROVED** [review]

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 2 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:113]: The comments explain the difference, but it's not clear from the message themselves:
- "parentShouldNotSummarize" is reported by the summarizer client (child client)
- "parentNoLongerSummarizer" is reported by the parent client

We could probably give these better names, or even unify them, but I didn't bother since the one reported by the summarizer client probably needs to be revisited anyway.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 3 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:495]: This is a good point.. I'll try to upgrade this to an error event.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 1 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:828]: What do you mean eliminate? We can make it a floating promise, but as it is now, this error should never hit already.

Comment (vladsud) [reply packages/runtime/container-runtime/src/summarizer.ts:828]: Take a look at handleSummaryAcks - there is pretty much no way it can throw exception. Essentially we have try/catch around try/catch and both report errors.
If it's expected that handleSummaryAcks is not propagating errors to callers, then .catch(() => {}) with some comment is better

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/summarizer.ts:113]: Looking at it in telemetry - what difference can I assume between the two? I.e. in both cases I'd guess it's because parent client disconnected at some point and is no longer oldest client, right?
I guess I'm not sure what kind of message are we sending trying to differentiate these cases. Maybe there is one, but it's not clear from these names / descriptions.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 1 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:828]: Okay I'm fine with that. I was thinking the telemetry log is slightly better, because it goes one of two ways:
1. As intended, it never gets hit, we never see this error in telemetry.
2. Not as intended, it gets hit (something in the underlying code changed!) and we log to telemetry + stop the .

Cons: Cost is this extra 3 lines of code to propagate the error and stop the summarizer client.

Pros: Benefit is no unhandled floating promise: linter is happy & we protect/log against future bugs by stopping the summarizer client. If we don't stop the summarizer client and somehow this gets hit, then this summarizer client is in a bad state since it can't update it's own state on summary ack; I think it would always do safe summary.

Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:828]: So to me this is more like "fail fast"/assert as it is now. Curious your thoughts.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 2 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:113]: Well it's just a "more information is better" type of thing. Which client realized it first? I'd expect it to be the parent client, but perhaps if something is wrong somewhere, we could help disambiguate it via this distinction.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 1 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/summarizer.ts:828]: I'd say I care only about presence (or lack of) handleSummaryAckFatalError in SummarizerStopReason - it feels like it should not be in enum as it should never be hit. Error in telemetry - sure.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 2 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/summarizer.ts:113]: That's fine, but it would be great to have some description on expectations (to help future OCEs realize what is normal and what is not and likely needs to look deeper).

For example, my naïve interpretations is something like that:
parentShouldNotSummarize should not happen that often because in majority of the cases we would first go through parentNotConnected and that would stop summarizer. But if reconnection happens quickly, we may find ourselves skipping parentNotConnected and stopping summarizer with parentShouldNotSummarize reason.
The difference between parentShouldNotSummarize and parentNoLongerSummarizer is roughly this ...

Basically I want to be able to map what is observed in telemetry to how code is expected to work and be able to spot anomalies (or earliest signals that something started to go wrong). It's rather hard to do as code is rather complicated, so any help in the form of inline documentation helps.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 4 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts]: I have changed this in the most recent commit.

**Comment by arinwt** [comment]
I have added "retry1" and "retry2" to be valid reasons for summarize attempts.

For all cases where we get perf end events in telemetry, we should be able to aggregate the counts by checking the new "Summarize_end" and "FailToSummarize" events like so:
```
count() "Summarize_end" where reason not starts with "retry" // count of successful first attempts
count() "Summarize_end" where reason == "retry1" // count of successful 2nd attempts
count() "Summarize_end" where reason == "retry2" // count of successful 3rd attempts
count() "FailToSummarize" // count of failed attempts
```

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 6]
```diff
@@ -840,19 +948,15 @@ export class Summarizer extends EventEmitter implements ISummarizer {
     }
 
     /** Implementation of SummarizerInternalsProvider.generateSummary */
-    public async generateSummary(options: {
-        fullTree: boolean,
-        refreshLatestAck: boolean,
-        summaryLogger: ITelemetryLogger,
-    }): Promise<GenerateSummaryData | undefined> {
+    public async generateSummary(options: IGenerateSummaryOptions): Promise<GenerateSummaryData> {
+        const result = this.internalsProvider.generateSummary(options);
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/summarizer.ts:952]: Note: Moving this _before_ the check to stop for "parentNoLongerSummarizer" allows for the last summary attempt to still go through before stopping, and still getting the generate summary results back to caller.

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 7]
```diff
@@ -100,17 +106,107 @@ export interface ISummarizerEvents extends IEvent {
...
+    /**
+     * Special case to attempt to summarize one last time before the
+     * summarizer client closes itself. This is to prevent cases where
+     * the summarizer client never gets a chance to summarize, because
+     * there are too many outstanding ops and/or parent client cannot
+     * stay connected long enough for summarizer client to catch up.
+     */
+    | "lastSummary"
+    /** Previous summary attempt failed, and we are retrying. */
+    | `retry${1 | 2}`;
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/summarizer.ts:183]: Note: Previously, we would reuse the same reason message that originally triggered the summary for the retries as well. Now it will replace it with this code; for example:
1. "maxOps"
2. "retry1"
3. "retry2"

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 8]
```diff
@@ -486,125 +594,125 @@ export class RunningSummarizer implements IDisposable {
...
+                refSequenceNumber: summaryOp.referenceSequenceNumber,
+                summarySequenceNumber: summaryOp.sequenceNumber,
+                handle: summaryOp.contents.handle,
+            });
 
-        return summaryData;
+            // Wait for ack/nack
+            const ackNack = await Promise.race([summary.waitAckNack(), pendingTimeoutP]);
+            if (!checkNotTimeout(ackNack)) {
+                return fail("summaryAckWaitTimeout", "Timeout while waiting for summaryAck/summaryNack op");
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/summarizer.ts:687]: Note: We will send 2 events and warning propagations for summary ack wait timeout now: 1 for the timer handler and 1 here. During the first load, only the ones for the timer handler will go off.

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 9]
```diff
@@ -486,125 +594,125 @@ export class RunningSummarizer implements IDisposable {
...
+                this.heuristics.ackLastSent();
+
+                // since we need a full summary after context reload, we only clear this on ack
+                this.immediateSummary = false;
+
+                summarizeEvent.end({ ...telemetryProps, handle: ackNack.contents.handle, message: "summaryAck" });
+                return true;
+            } else {
+                return fail(
+                    "summaryNack",
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/summarizer.ts:707]: Note: We no longer emit "SummaryAck" or "SummaryNack" events now. Instead we send "Summarize_end" with message "summaryAck", and "Summarize_cancel" with message "summaryNack".

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 10]
```diff
@@ -486,125 +594,125 @@ export class RunningSummarizer implements IDisposable {
...
-        }
-    }
-
-    private async generateSummaryWithLogging(message: string, safe: boolean): Promise<GenerateSummaryData | undefined> {
-        const summarizingEvent = PerformanceEvent.start(this.logger, {
-            eventName: "GenerateSummary",
-            message,
+        const summarizeEvent = PerformanceEvent.start(this.logger, {
+            eventName: "Summarize",
+            reason,
```
Comment (arinwt) [inline packages/runtime/container-runtime/src/summarizer.ts:609]: This code has been renamed from `message` to `reason`.

Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:609]: Now `message` represents the failure/success message.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 11]
```diff
@@ -1661,48 +1661,66 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
             await this.deltaManager.inbound.pause();
 
-            const attemptData: Omit<IUnsubmittedSummaryData, "reason"> = {
+            const summaryRefSeqNum = this.deltaManager.lastSequenceNumber;
+            const message = `Summary @${summaryRefSeqNum}:${this.deltaManager.minimumSequenceNumber}`;
+
+            this.summarizerNode.startSummary(summaryRefSeqNum, summaryLogger);
+
+            // Helper function to check whether we should still continue between each async step.
+            const checkContinue = (): { continue: true; } | { continue: false; error: string } => {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1673]: Fills like `error: string | undefined` is a bit easier to work with

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 12]
```diff
@@ -1661,48 +1661,66 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
             const generateData: IGeneratedSummaryData = {
                 summaryStats: summarizeResult.stats,
                 generateDuration: trace.trace().duration,
             };
 
-            if (!this.connected) {
-                return { ...attemptData, ...generateData, reason: "disconnected" };
+            continueResult = checkContinue();
+            if (!continueResult.continue) {
+                return { ...attemptData, ...generateData, error: continueResult.error };
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1721]: Signature of the function is not conveying that it returns extra props (IGeneratedSummaryData) when submitted === false.

Comment (vladsud) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1721]: There are more props a bit below:
  return { ...attemptData, ...generateData, ...uploadData, error: continueResult.error };

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 13]
```diff
@@ -1717,9 +1735,12 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
                 };
 
-            const handle = await this.storage.uploadSummaryWithContext(
-                summarizeResult.summary,
-                summaryContext);
+            let handle: string;
+            try {
+                handle = await this.storage.uploadSummaryWithContext(summarizeResult.summary, summaryContext);
+            } catch (error) {
+                return { ...attemptData, ...generateData, error };
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1742]: I wonder if there is better way to deal with errors. I.e. have caller pass in object on which this function sets properties as they become available (all props are optional). And then remove all try/catch like this, leaving it up to the caller to have single try/catch, examine that object with optional props and do logging.

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 14]
```diff
@@ -100,17 +106,107 @@ export interface ISummarizerEvents extends IEvent {
...
+     * op sequence numbers with the reference sequence number
+     * of the summarize op corresponding to the last summary
+     * ack op.
+     */
+    | "maxOps"
+    /**
+     * Special case to generate a summary immediately after starting.
+     * @deprecated - only used in context reload hot swap, which was removed
+     */
+    | "immediate"
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:168]: Why not remove deprecated items from this list? It does not seem like there are any references to it in the code.
Similar for save op, though if we leave it, can you add prefix to the message indicating it's because of save op?

File: packages/runtime/container-runtime/src/summarizer.ts [hunk 15]
```diff
@@ -479,121 +587,121 @@ export class RunningSummarizer implements IDisposable {
...
+            refreshLatestAck,
+            fullTree,
             timeSinceLastAttempt: Date.now() - this.heuristics.lastAttempted.summaryTime,
             timeSinceLastSummary: Date.now() - this.heuristics.lastAcked.summaryTime,
-            safe: safe || undefined,
         });
+        // Helper function to report failures and return.
+        const fail = <K extends keyof ISummarizeError>(
+            message: K,
+            description: ISummarizeError[K],
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/summarizer.ts:611]: description strings will be present twice in the code, adding to bundle size. Can this function retrieve description from ISummarizeError? Maybe by changing ISummarizeError shape (it's essentially a map).

Comment (vladsud) [reply packages/runtime/container-runtime/src/summarizer.ts:611]: Actually, can we get rid of strings (description) altogether? English strings are not useful anyway (other than for telemetry), clients are much better to work with IDs that they can map to their localized strings if needed. raiseSummarizingError() ideally should either have enum arg, or some kind of structured error (ideally relying on same structure as everywhere else - defining its on subset of errorType(s)

**Comment by vladsud** [comment]
It feels like this PR is way more than original topic :) Might be worth updating title & description

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 15 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:611]: `ISummarizeError` as an interface should only exists before being transpiled and bundled into js code. It will appear 2x in source code/typings only.

Do you think I should promote `ISummarizeError` to be a readonly const object: `summarizeErrors`, acting as a map from code to description?

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1742]: I thought about doing this... it would go hand-in-hand with the potential feature of adding properties along the way to performance events within execute/executeAsync.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 14 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:168]: Added "save;" prefix. And removed "immediate". I think a separate PR could probably just fully remove save ops from the system.

File: packages/runtime/container-runtime/src/summarizer.ts (hunk 15 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/summarizer.ts:611]: I'm not sure descriptions are valuable / useful at all. What's the scenario? They might be useful only in telemetry (as they are not localized), and I think code are as good there.

Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizer.ts:611]: Went ahead and converted `ISummarizeError` interface to `summarizeErrors` const object. We now use that to map from code/key to description.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 13 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1742]: I looked at doing this a little bit- it will make the code slimmer, so I think probably it's worth it, but I want to punt it out of this PR. It's too big and that has additional changes to keep the strict typings.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 12 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1721]: I think it does...
It returns either `IUnsubmittedSummaryData` or `ISubmittedSummaryData`, which are distinguishable by `submitted` as `false` or `true` respectively.

If `submitted === false`, then we can see `IUnsubmittedSummaryData` actually extends partials of both `IGeneratedSummaryData` and `IUploadedSUmmaryData`.

Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1721]: Agree this isn't the easiest to read set of types, and will probably simplify them when using parameter object instead of return type (from other comment). But you can see definition of `IUnsubmittedSummaryData`:
```ts
export interface IUnsubmittedSummaryData extends Partial<IGeneratedSummaryData>, Partial<IUploadedSummaryData> { ... }
````

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 11 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1673]: Pushing this out until using parameter object instead of giant return object, because it will probably change this too.
